### PR TITLE
Add `puncons`, `punsafeUncons` in `Plutarch.List`

### PR DESCRIPTION
**Pull request by Renegatto**

### CODE REVIEW

File: Plutarch/List.hs [hunk 1]
```diff
@@ -133,6 +138,20 @@ pconvertLists = phoistAcyclic $
       (\x xs -> pcons # x #$ self # xs)
       pnil
 
+-- | Extract head and tail of the list, throws error if list is empty.
+punsafeUncons ::
```
Comment (L-as) [inline Plutarch/List.hs:142]: Note, this isn't unsafe. This should be named `ptryUncons` or similar.

File: Plutarch/List.hs [hunk 2]
```diff
@@ -133,6 +138,20 @@ pconvertLists = phoistAcyclic $
       (\x xs -> pcons # x #$ self # xs)
       pnil
 
+-- | Extract head and tail of the list, throws error if list is empty.
+punsafeUncons :: 
+  PIsListLike list a =>
+  Term s (list a) ->
+  Term s (PPair a (list a))
+punsafeUncons = pelimList (\x -> pcon . PPair x) perror
```
Comment (L-as) [inline Plutarch/List.hs:146]: This should also be hoisted and a Plutarch-level function.

File: Plutarch/List.hs [hunk 3]
```diff
@@ -67,6 +71,7 @@ import Plutarch.Bool (PBool (PFalse, PTrue), PEq, pif, (#&&), (#==), (#||))
 import Plutarch.Integer (PInteger)
 import Plutarch.Lift (pconstant)
 import Plutarch.Pair (PPair (PPair))
+import Plutarch.Maybe (PMaybe (PJust,PNothing))
```
Comment (L-as) [inline Plutarch/List.hs:74]: `./bin/format`

**Review by L-as — CHANGES_REQUESTED** [review]
Thanks for the contribution!

File: Plutarch/List.hs [hunk 4]
```diff
@@ -133,6 +138,20 @@ pconvertLists = phoistAcyclic $
...
 
+-- | Extract head and tail of the list, throws error if list is empty.
+punsafeUncons :: 
+  PIsListLike list a =>
+  Term s (list a) ->
+  Term s (PPair a (list a))
+punsafeUncons = pelimList (\x -> pcon . PPair x) perror
+
+-- | Extract head and tail of the list, if list is not empty.
+puncons ::
```
[the remark this answers is not in the document]
Comment (Renegatto) [reply Plutarch/List.hs]: What exactly to hoist here since it's not recursive?

File: Plutarch/List.hs (hunk 2 above)
Comment (Renegatto) [reply Plutarch/List.hs]: What exactly to hoist here?

**Comment by L-as** [comment]
Hoisting has nothing to do with recursion. It's to do CSE. You hoist closed terms generally.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,7 @@
 # Revision history for plutarch
 
 # Unreleased
+- Add `puncons` and `punsafeUncons` functions for deconstructing list.
```
Comment (L-as) [inline CHANGELOG.md:4]: We should fix this in another PR for fixing the changelog

**Review by L-as — APPROVED** [review]
LGTM

**Comment by srid** [comment]
@L-as What do you think of requiring new PRs like this one that adds new code to also add the corresponding examples to `plutarch-test` (so we get uplc/bench goldens, to test for any potential regression in the future)?

**Pull request merged**